manubot/pandoc/cite_filter.py

```diff
@@ -0,0 +1,219 @@
+  --filter=pandoc-manubot-cite \
+  --filter pandoc-citeproc \
+  manubot/pandoc/tests/test_cite_filter/input.md
+
+# call the filter manually using pandoc JSON output
+pandoc \
+  --to=json \
+  manubot/pandoc/tests/test_cite_filter/input.md \
+  | python manubot/pandoc/test_cite.py markdown
+```
+
+Related resources on pandoc filters:
+
+- [python pandocfilters package](https://github.com/jgm/pandocfilters)
+- [python panflute package](https://github.com/sergiocorreia/panflute)
+- [panflute Citation class](http://scorreia.com/software/panflute/code.html#panflute.elements.Citation)
+"""
+import argparse
+import logging
+import pathlib
+
+import panflute as pf
+
+from manubot.cite.citekey import is_valid_citekey
+from manubot.process.util import (
+    get_citekeys_df,
+    generate_csl_items,
+    load_manual_references,
+    write_csl_json,
+)
+
+
+global_variables = {
+    "manuscript_citekeys": list(),
+}
+
+
+def parse_args():
+    """
+    Read command line arguments
+    """
+    parser = argparse.ArgumentParser(
+        description="Pandoc filter for citation by persistent identifier. "
+        "Filters are command-line programs that read and write a JSON-encoded abstract syntax tree for Pandoc. "
+        "Unless you are debugging, run this filter as part of a pandoc command by specifying --filter=pandoc-manubot-cite."
+    )
+    parser.add_argument("target_format")
+    parser.add_argument(
+        "--input",
+        nargs="?",
+        type=argparse.FileType("r", encoding="utf-8"),
+        help="path read JSON input (defaults to stdin)",
+    )
+    parser.add_argument(
+        "--output",
+        nargs="?",
+        type=argparse.FileType("w", encoding="utf-8"),
+        help="path to write JSON output (defaults to stdout)",
+    )
+    args = parser.parse_args()
+    return args
+
+
+def _get_citekeys_action(elem, doc):
+    """
+    Panflute action to extract citationId from all Citations in the AST.
+    """
+    if not isinstance(elem, pf.Citation):
+        return None
+    manuscript_citekeys = global_variables["manuscript_citekeys"]
+    manuscript_citekeys.append(elem.id)
+    return None
+
+
+def _citation_to_id_action(elem, doc):
+    """
+    Panflute action to update the citationId of Citations in the AST
+    with their manubot-created keys.
+    """
+    if not isinstance(elem, pf.Citation):
+        return None
+    mapper = global_variables["citekey_shortener"]
+    if elem.id in mapper:
+        elem.id = mapper[elem.id]
+    return None
+
+
+def _get_reference_link_citekey_aliases(elem, doc):
+    """
+    Based on TypeScript implementation by phiresky at
+    https://github.com/phiresky/pandoc-url2cite/blob/b28374a9a037a5ce1747b8567160d8dffd64177e/index.ts#L118-L152
+
+    Uses markdown's link reference syntax to define citekey aliases (tags)
+    https://spec.commonmark.org/0.29/#link-reference-definitions
+    """
+    if type(elem) != pf.Para:
+        return
+    while (
```

Comment:
It's hard to follow this loop without knowing the structure of `pf.Para`.  Is this applied only on Markdown link references?  Or it's extracting citekeys and their aliases from the entire document?

Reply:
comments added.

> Is this applied only on Markdown link references? Or it's extracting citekeys and their aliases from the entire document?

The former